### PR TITLE
Make django-lifecycle much, much faster

### PR DESCRIPTION
**Pull request by dralley**
Some of the work that the lifecycle mixin is during the initialization of new model objects is very expensive and unnecessary. It's calculating (and caching) field names and foreign key models per-object, rather than per-class / model.  All instances of a model are going to have the same field names and foreign key model types so this work actually only needs to be done once per model type.

Replacing cached methods with cached classmethods yields a *very* sizable performance improvement when creating a bunch of new model instances.

### CODE REVIEW

**Comment by dralley** [comment]
We have code that saves and deletes a lot of models (tens of thousands) in a loop.  It got about 40% faster with these changes.

Before:

before

After:

after

So you can see that it actually makes a pretty huge difference.

**Comment by dralley** [comment]
Our code creates and saves a bunch of models, and it got about 40% faster after applying this patch.  I did some profiling and our application went from spending ~40% of its compute time inside of django-lifecycle to a fraction of one percent.

Before:

before

After:

after

**Comment by dralley** [comment]
I didn't measure this but I have to imagine that the deduplicated caches would save a lot of memory as well.

**Comment by dralley** [comment]
Here are some profiles from a completely different part of our codebase, where it's just deleting a lot of existing models.  The time taken went from ~214 seconds to ~77 seconds.

## Before:

orphan_before

## After:

orphan_after

**Comment by dralley** [comment]
I finally got the tests passing.  The changes necessary to do so are in a separate commit in case it's easier to review that way.

The performance difference is pretty significant. Our code does a lot of heavy creating, saving, and deleting of model instances, tens of thousands of them in a tight loop, and our speedup ranges from 40% to 65%.

Here's some cProfile graphs showing before and after for different parts of our code.

------------

## Before:

before

## After:

after

-------------

## Before:

orphan_before

## After:

orphan_after

**Comment by dralley** [comment]
(The graphs don't show every single function called, just the ones significant enough impact to be worth paying attention to.  That's why the other blocks disappeared -- they're still being called, but only like <10 times instead of hundreds of thousands of times like before).

File: django_lifecycle/mixins.py [hunk 1]
```diff
@@ -140,42 +141,44 @@ def delete(self, *args, **kwargs):
...
 
-        for name in dir(self):
+        for name in dir(cls):
             if name in skip:
                 continue
             try:
-                attr = getattr(self, name)
-                if ismethod(attr) and hasattr(attr, "_hooked"):
+                attr = getattr(cls, name)
+                if isfunction(attr) and hasattr(attr, "_hooked"):
```
Comment (dralley) [inline django_lifecycle/mixins.py:154]: Minor semantic change right here - I think the type of the method/function under the hood depends on whether you're looking at it from the instance perspective (where methods are special) or from the class perspective (where methods are just functions and there's no magic going on).

I don't think it's problematic but I want to call it out for a little extra attention.

File: django_lifecycle/mixins.py [hunk 2]
```diff
@@ -196,15 +199,15 @@ def _run_hooked_methods(self, hook: str) -> List[str]:
                 if when_field:
                     if self._check_callback_conditions(when_field, callback_specs):
                         fired.append(method.__name__)
-                        method()
+                        method(self)
```
Comment (dralley) [inline django_lifecycle/mixins.py:202]: The flip side of that thing I mentioned earlier, it's returning functions not bound methods, so we can't rely on internal Python magic to pass ```self```.  But functionally it's the same thing.

**Comment by dralley** [comment]
I can squash the commits once you're done reviewing also

**Comment by rsinger86** [comment]
This is great. Thank you very much.

**Pull request merged**